### PR TITLE
configuration: fix the '%%' escape in configuration values

### PR DESCRIPTION
**Pull request by pdgendt**
configparser interpolates values when they are read, so a literal '%'
has to be written as '%%'. That escape has not worked since v0.13.0:
`_copy_to_configparser()` re-sets the already interpolated value into a
second interpolating parser, which rejects the '%' it just produced.
Because `WestApp.run()` copies every option before dispatching a command,
a single escaped value makes every west command fail:

```shell
$ cat .west/config
[manifest]
path = zephyr
[build]
cmake-args = -DFOO=100%%

$ west topdir
ValueError: invalid interpolation syntax in '-DFOO=100%' at position 12
```

Escape the value again on the way into the compatibility object, which
interpolates it back to what `Configuration.get()` returns.

With the escape working, make it unnecessary to know about: 'west config'
now escapes what it writes, so a value containing a '%' can be set from
the command line and is read back verbatim. Only hand-written files need
the escape, which is what the module documentation now says.

An unescaped '%' remains an error, but no longer a configparser traceback
from an unrelated command: interpolation failures are reported as
`MalformedConfig` naming the option and how to fix it, like the other
malformed configuration errors.

Interpolation itself keeps working, so '%(other-option)s' references in
existing configuration files are unaffected.

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -910,3 +910,53 @@ def sorted_list(other_args=''):
 def test_round_trip():
     cmd('config pytest.foo bar,baz')
     assert cmd('config pytest.foo').strip() == 'bar,baz'
+
+
+def test_round_trip_percent():
+    # A '%' does not have to be avoided on the command line: it is
```
Comment (marc-hb) [inline tests/test_config.py]:
```suggestion
    # A '%' does not have to be manually escaped on the command line: it is automatically
```

File: src/west/configuration.py [hunk 2]
```diff
@@ -398,8 +426,10 @@ def load(cf: _InternalCF):
                     continue
                 if section not in cp:
                     cp.add_section(section)
-                for key, value in contents.items():
-                    cp[section][key] = value
+                for key in contents:
+                    # 'cp' interpolates when it is read from, so the
+                    # value goes in escaped: it is interpolated already.
```
Comment (marc-hb) [inline src/west/configuration.py]:
```suggestion
                    # value must be re-escaped.
```

File: src/west/configuration.py [hunk 3]
```diff
@@ -55,6 +60,13 @@ def _configparser():  # for internal use
     return configparser.ConfigParser(allow_no_value=True)
 
 
+def _escape(value):  # for internal use
```
Comment (marc-hb) [inline src/west/configuration.py]:
```suggestion
def _escape_percent(value):  # for internal use
```

File: src/west/configuration.py [hunk 4]
```diff
@@ -112,13 +124,29 @@ def getint(self, option: str):
...
         return self._get(option, self.cp.getfloat)
 
+    def value(self, section: str, key: str):
+        # The interpolated value of an option, by section and key.
+        try:
+            return self.cp.get(section, key)
+        except configparser.InterpolationError as err:
+            raise self._malformed_value(section, key) from err
+
+    def _malformed_value(self, section: str, key: str) -> MalformedConfig:
```
Comment (marc-hb) [inline src/west/configuration.py:134]: Move this before first use?

I generally don't care much about the order in which functions are defined but here it looks weird for this to be used before AND after its definition; I feel like this is crossing a line. Before or after: pick one? :-)

**Review by marc-hb — COMMENTED** [review]
Generally happy with the approach but to be sure this is sound I would very much like to see test coverage for more edge cases with triple and quadruple `%`, both from the command line and from the API.

File: src/west/configuration.py [hunk 5]
```diff
@@ -33,6 +33,11 @@
 Configuration values from later configuration files override configuration
 from earlier ones. Local values have highest precedence, and system values
 lowest.
+
+Values are interpolated when they are read: a literal ``%`` must be
```
Comment (marc-hb) [inline src/west/configuration.py]:
```suggestion
https://docs.python.org/3/library/configparser.html interpolates values when they are read
```
I meant to and forgot to ask for this in the last round, sorry.

Please phrase it better and the exact URL may not even be necessary but I think it's important to mention that we defer to this standard Python module so:
- There is a pointer to the exhaustive documentation in case anyone needs it.
- In case of issues with the configparser itself like #779 (now fixed in configparser), people realize that the `configparser` version / Python version matters.

File: src/west/configuration.py [hunk 6]
```diff
@@ -33,6 +33,11 @@
...
+written as ``%%``, and ``%(other-option)s`` expands to the value of
+another option from the same section. Values written by ``west config``
+are escaped automatically, so only hand-written files need care.
```
Comment (marc-hb) [inline src/west/configuration.py]: I think we should add that interpolation features cannot be used from the command line, nor from the West APIs, correct? It requires manual config file edition or the Python configparser API.

Fine by me, it just needs to be stated.

File: tests/test_config.py [hunk 7]
```diff
@@ -910,3 +910,89 @@ def sorted_list(other_args=''):
...
+    cmd(['config', 'pytest.v', value])
+
+    assert cmd('config pytest.v').strip() == value
+    escaped = value.replace('%', '%%')
+    assert f'v = {escaped}\n' in local_config().read_text()
+
+
+@pytest.mark.parametrize('value', PERCENT_VALUES)
+def test_round_trip_percent_api(value):
+    # Same through the API, which is what extension commands use.
```
Comment (marc-hb) [inline tests/test_config.py]:
```suggestion
    # Same through the west API, which is what extension commands use.
```
... as opposed to the configparser API. Admittedly low risk of confusion but this is just one extra word.

**Review by marc-hb — COMMENTED** [review]
Just a few, small doc and comments changes. Everything else LGTM!

File: src/west/configuration.py (hunk 6 above)
Comment (marc-hb) [reply src/west/configuration.py]: > interpolation features cannot be used from the command line, nor from the West APIs, correct? 

You could argue that this is very strictly speaking a regression...

But I think it's fine considering:
- The interpolation feature was never advertised.
- It's becoming even more of a stretch to imagine someone using interpolation AND scripting config changes at the same time!
- Unlike the abandoned #996, the interpolation feature is still available and usable. It only requires some code changes to keep using it.

**Review by marc-hb — APPROVED** [review]

**Review by kartben — APPROVED** [review]
